sys/kern/mcount.c

```diff
@@ -0,0 +1,187 @@
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*-
+ * Copyright (c) 1983, 1992, 1993
+ *	The Regents of the University of California.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the University nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/types.h>
+#include <sys/gmon.h>
+#include <sys/interrupt.h>
+#include <sys/spinlock.h>
+#include <sys/types.h>
+
+static SPIN_DEFINE(mcount_lock, 0);
+
+void __cyg_profile_func_enter(void *from, void *self) {
+  u_long frompc = (u_long)from, selfpc = (u_long)self;
+  u_short *frompcindex;
+  tostruct_t *top, *prevtop;
+  gmonparam_t *p = &_gmonparam;
+  long toindex;
+
+  /*
+   * check that we are profiling
+   * and that we aren't recursively invoked.
+   */
+  if (p->state != GMON_PROF_ON)
+    return;
+
+  intr_disable();
+
+
+  WITH_SPIN_LOCK (&mcount_lock) {
+    p->state = GMON_PROF_BUSY;
+    /*
+    * check that frompcindex is a reasonable pc value.
+    * for example:	signal catchers get called from the stack,
+    *		not from text space.  too bad.
+    */
+    frompc -= p->lowpc;
+    if (frompc > p->textsize)
+      goto done;
+
+    size_t index = (frompc / (p->hashfraction * sizeof(*p->froms)));
```

Comment:
If you use `HASHFRACTION` here instead of `p->hashfraction` then then the compiler will probably optimize the division into a shift, at least with `-O2`.

Reply:
Sure!